renames transform method in base classes to check_transform_input_and_state

solegalli opened the pull request:
fixes #670

solegalli commented:
@ClaudioSalvatoreArcidiacono FYI

Would you like to have a look? All I did is rename the method in the base classes.

ClaudioSalvatoreArcidiacono commented:
Hey @solegalli, it would be a pleasure to help reviewing this PR!

ClaudioSalvatoreArcidiacono commented:
Mh... I see that there are still some base classes that have not been updated, namely:

```
6 results - 6 files

feature_engine/discretisation/base_discretiser.py:
  42  
  43:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
  44          """Sort the variable values into the intervals.

feature_engine/encoding/base_encoder.py:
  191  
  192:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
  193          """Replace categories with the learned parameters.

feature_engine/imputation/base_imputer.py:
  44  
  45:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
  46          """

feature_engine/outliers/base_outlier.py:
  65  
  66:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
  67          """

feature_engine/selection/base_selector.py:
  66  
  67:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
  68          """

feature_engine/timeseries/forecasting/base_forecast_transformers.py:
  167  
  168:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
  169          """
```

is it intentional?

ClaudioSalvatoreArcidiacono commented:
> I was wondering if there was a better way of solving this problem... like, why does scikit-learn need to re-convert the output of the transformer to pandas if it is already in pandas? Is it maybe needed to set something in the transformer to tell scikit-learn that the output of the transformer is always pandas? perhaps some output configs settings?

I see that this is not possible, sometimes columns in sklearn pipelines are renamed with the `<step_name>.<col_name>` convention, so even if the output of the estimator is a data frame, sklearn still needs to rename the columns of the data frame.

solegalli commented:
Yeah, I looked at their source code and I also got the impression that it couldn't be done.

solegalli commented:
> Mh... I see that there are still some base classes that have not been updated, namely:
> 
> ```python
> # 6 results - 6 files
> 
> # feature_engine/discretisation/base_discretiser.py:
>   43:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
> 
> # feature_engine/encoding/base_encoder.py:
>   192:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
> 
> # feature_engine/imputation/base_imputer.py:
>   45:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
> 
> # feature_engine/outliers/base_outlier.py:
>   66:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
> 
> # feature_engine/selection/base_selector.py:
>   67:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
> 
> # feature_engine/timeseries/forecasting/base_forecast_transformers.py:
>   168:     def transform(self, X: pd.DataFrame) -> pd.DataFrame:
> ```
> 
> is it intentional?
> 
> For `BaseDiscretizer` it is not needed as it extends `BaseNumericalTransformer` but I think others like `CategoricalMethodsMixin` should be updated

Thanks. I added tests for all transformers now, and changed the methods in the forecasting classes. Should be OK now.

I just missed them, there are too many, lol.

Thanks for flagging this.

solegalli commented:
If test passes, I didn't change the class. Some, like BaseOutlier, the transform method IS the transform method. Not an intermediary step like in the other base classes.

ClaudioSalvatoreArcidiacono commented on tests/test_sklearn_compatible/test_set_output.py:
```diff
@@ -1,5 +1,6 @@
 import numpy as np
 import pandas as pd
+
```
```suggestion
```

ClaudioSalvatoreArcidiacono commented on tests/test_selection/test_information_value.py:
```diff
@@ -1,4 +1,5 @@
 import math
+
```
Maybe a good idea to run isort on these files as well?

ClaudioSalvatoreArcidiacono commented on tests/test_outliers/test_check_estimator_outliers.py:
```diff
@@ -21,3 +23,19 @@ def test_check_estimator_from_feature_engine(estimator):
+
+
+@pytest.mark.parametrize("transformer", _estimators)
+def test_transformers_in_pipeline_with_set_output_pandas(transformer):
```
This test will fail when

```python
transformer = Winsorizer(add_indicators=True)
```

solegalli replied:
Thank you! I missed it.

I had to make a workaround that doesn't make me happy to modify the BaseOutlier.

I need to look at class inheritance at some point. It's getting out of hand.

Thanks a lot for flagging this one!

ClaudioSalvatoreArcidiacono commented:
Hey @solegalli, I think I am done reviewing :)

I have left a couple of comments.

Again, thanks a lot for all of the passion you put in this library!

solegalli commented:
Thanks a lot for the review @ClaudioSalvatoreArcidiacono I really appreciate it!!!
